### ISSUE TITLE
rtkit: doesn't work on musl

### ISSUE THREAD

**Issue by st3r4g** [void-linux/void-packages#28728]
`rtkit` is basically a dbus interface to `sched_setscheduler`, which is not implemented on musl.

Here is the reason why:
https://github.com/MusicPlayerDaemon/MPD/issues/218#issuecomment-364320830

So it seems it should be patched to use the `pthread_*` functions instead.

**Comment by ericonr** [comment]
We might have to call the syscalls directly, I think. The functions it uses take a PID parameter, while pthread ones only allow you to act on your own process...

From the comments there, does this even work on glibc properly? Or does it end up applying the parameter to a single thread in the process?

**Comment by Logarithmus** [comment]
It seems that the problem is both with `musl` & missing `CONFIG_RT_GROUP_SCHED` flag.

> Currently I'm investigating these stability issues with `pipewire`. It seems that `pipewire` is trying to use `rtkit` to make audio/video processing run in real time. Unfortunately, it fails to do so:
> ```
> 2020-11-16T03:49:59.73350 daemon.err: Nov 16 03:49:59 rtkit-daemon[3165]: Failed to make ourselves RT: Function not implemented
> 2020-11-16T03:49:59.73602 daemon.warn: Nov 16 03:49:59 rtkit-daemon[3165]: Warning: failed to read scheduler policy: Function not implemented
> 2020-11-16T03:49:59.73620 daemon.debug: Nov 16 03:49:59 rtkit-daemon[3165]: Supervising 0 threads of 0 processes of 1 users.
> 2020-11-16T03:49:59.73645 daemon.debug: Nov 16 03:49:59 rtkit-daemon[3165]: Supervising 0 threads of 0 processes of 1 users.
> 2020-11-16T03:49:59.74118 daemon.err: Nov 16 03:49:59 rtkit-daemon[3165]: Failed to make ourselves RT: Function not implemented
> ```
> After a bit of searching, I've found these bugs:
> https://bugzilla.redhat.com/show_bug.cgi?id=1229700
> https://bugs.gentoo.org/show_bug.cgi?id=569546
>
> I've built `linux5.9` for `x86_64` with `CONFIG_RT_GROUP_SCHED` disabled.
> For some strange reason, nothing changed, the error persists. Luckily, screen sharing works without `rtkit`.
> I guess that those rare issues with screen sharing, which were described above, may be caused by `rtkit` not working.

**Comment by st3r4g** [comment]
Well this is a different issue
>missing CONFIG_RT_GROUP_SCHED kernel option.

you mean enabled right? void has it enabled.

I've seen the warning against `CONFIG_RT_GROUP_SCHED` in the [jack faq](https://jackaudio.org/faq/linux_group_sched.html) but haven't experienced such a problem, both with `jack` on `x86_64-musl` and `pulseaudio`/`pipewire` on `x86_64`.

I'm letting `rtkit` start via dbus activation instead of manually enabling the service. Mayeb this could be the reason?

**Comment by Logarithmus** [comment]
@st3r4g yes, I mean that all major distros have this option disabled (because it interfers with `pulseaudio`, `jack` & `pipewire` in some way).
5 years ago Lennart Pottering himself asked it to be disabled in Fedora, lol: https://bugzilla.redhat.com/show_bug.cgi?id=1229700
But Void has it enabled. Some months ago I've built my own kernel without `CONFIG_RT_GROUP_SCHED`, but the error/warning persisted. I was testing on `x86_64-musl`.

**Comment by st3r4g** [comment]
> but the error/warning persisted I was testing on `x86_64-musl`.

then `CONFIG_RT_GROUP_SCHED` is not the problem... it's the fact you were on musl.

**Comment by Logarithmus** [comment]
OK, then I don't understand why did Lennart asked to disable it? And why do other distros have it disabled too?

On February 21, 2021 4:26:11 PM GMT+03:00, Stefano Ragni <notifications@github.com> wrote:
>> but the error/warning persisted I was testing on `x86_64-musl`.
>
>then `CONFIG_RT_GROUP_SCHED` is not the problem... it's the fact you
>were on musl.

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by 12101111** [comment]
Enable `CONFIG_RT_GROUP_SCHED` make rtkit don't work either: https://bugs.gentoo.org/569546

**Comment by st3r4g** [comment]
> Enable `CONFIG_RT_GROUP_SCHED` make rtkit don't work either: https://bugs.gentoo.org/569546

Void has it enabled:
```
[void-packages 0]$ grep CONFIG_RT_GROUP_SCHED srcpkgs/linux5.10/files/x86_64-dotconfig 
CONFIG_RT_GROUP_SCHED=y
```
and for me rtkit works (on glibc):
```
[void-packages 0]$ /usr/libexec/installed-tests/rtkit/rtkit-test 
Max realtime priority is: 20
Min nice level is: -15
Rttime limit is: 200000 ns
before:
        SCHED_RESET_ON_FORK: no
        SCHED_OTHER with nice level: 0
Successfully became high priority.
after high priority:
        SCHED_RESET_ON_FORK: yes
        SCHED_OTHER with nice level: -10
Successfully became realtime.
after realtime:
        SCHED_RESET_ON_FORK: yes
        SCHED_RR with priority 10
```
so I don't know...

**Comment by yemouu** [comment]
This seems to be the [approach](https://gitlab.alpinelinux.org/alpine/aports/-/blob/f7f0e8106bd5fcacd5384c50187c1159f11fc9eb/testing/rtkit/sched_getscheduler_sched_setscheduler.patch) that Alpine will take

**Comment by ericonr** [comment]
I believe that patch isn't completely right, it's using a `static struct sched_params def` instead of the `params` argument, which makes `rtkit` basically a no-op. Otherwise yes, the approach makes sense (though I would have `#define sched_setscheduler my_sched_setscheduler` instead ;)

**Comment by ericonr** [comment]
Alpine's patch has been fixed, I'm looking into importing it to void.

**Comment by ericonr** [comment]
It seems to only be available for the system pipewire service, not for the user one, since it uses capabilities as set by systemd.

If we create a system service for pipewire we could set this up, though. Yggdrasil uses capabilities in its service as well.
